plugins/PRAplugin/src/MCSSolver.py

```diff
@@ -0,0 +1,146 @@
+    container.invMapping = {}
+
+    for child in xmlNode:
+      if child.tag == 'topEventID':
+        container.topEventID = child.text.strip()
+      elif child.tag == 'solverOrder':
+        try:
+          self.solverOrder = int(child.text.strip())
+        except:
+          raise IOError("MCSSolver: xml node solverOrder must contain an integer value")
+      elif child.tag == 'variables':
+        variables = [str(var.strip()) for var in child.text.split(",")]
+      elif child.tag == 'map':
+        container.mapping[child.get('var')]      = child.text.strip()
+        container.invMapping[child.text.strip()] = child.get('var')
+      else:
+        raise IOError("MCSSolver: xml node " + str(child.tag) + " is not allowed")
+
+  def initialize(self, container, runInfoDict, inputFiles):
+    """
+      Method to initialize this plugin
+      @ In, container, object, self-like object where all the variables can be stored
+      @ In, runInfoDict, dict, dictionary containing all the RunInfo parameters (XML node <RunInfo>)
+      @ In, inputFiles, list, list of input files (if any)
+      @ Out, None
+    """
+    pass
+
+  def createNewInput(self, container, inputs, samplerType, **Kwargs):
+    """
+      This function has been added for this model in order to be able to create a FTstructure from multiple files
+      @ In, container, object, self-like object where all the variables can be stored
+      @ In, inputs, list, the inputs (list) to start from to generate the new one
+      @ In, samplerType, string, is the type of sampler that is calling to generate a new input
+      @ In, **kwargs, dict,  is a dictionary that contains the information coming from the sampler,
+           a mandatory key is the sampledVars'that contains a dictionary {'name variable':value}
+      @ Out, ([(inputDict)],copy.deepcopy(kwargs)), tuple, return the new input in a tuple form
+    """
+    if len(inputs) > 1:
+      raise IOError("MCSSolver: More than one file has been passed to the MCS solver")
+
+    mcsIDs, probability, mcsList, beList = mcsReader(inputs[0])
+
+    self.topEventTerms = {}
+
+    # mcsList is supposed to be a list of lists
+    # E.g., if the MCS are ABC CD and AE --> MCS1=['A','B','C'], MCS2=['D','C'], MCS3=['A','E']
+    #       then mcsList = [MCS1,MCS2,MCS3] =
+    #                    = [['A', 'B', 'C'], ['D', 'C'], ['A', 'E']]
+    # Top event should be:   ABC + CD + AE +
+    #                      - ABCD - ABCE - ACDE
+    #                      + ABCDE
+
+    for order in range(1,self.solverOrder+1):
+      self.topEventTerms[order]=[]
+      terms = list(itertools.combinations(mcsList,order))
+      # terms is a list of tuples
+      # E.g., for order=2: [ (['A', 'B', 'C'], ['D', 'C']),
+      #                      (['A', 'B', 'C'], ['A', 'E']),
+      #                      (['D', 'C'], ['A', 'E']) ]
+      self.topEventTerms[order] = []
+      for term in terms:
+        basicEventCombined = set(itertools.chain(*term))
```

Comment:
without the if check, you can do a list comprehensions for the for loop.

Reply:
also change itertools.chain to itertools.chain.from_iterable, see post: https://stackoverflow.com/a/30774051
```
>>> timeit.timeit("[list(set(reduce ( list.__add__, big)))]", setup="big = [ [10,20,30,40] for ele in range(10000)]", number=30)
16.051744650801993
>>> timeit.timeit("[list(set(reduce ( itertools.chain, big)))]", setup="big = [ [10,20,30,40] for ele in range(10000)]", number=30)
54.721315866467194
>>> timeit.timeit("list(set(itertools.chain.from_iterable(big)))", setup="big = [ [10,20,30,40] for ele in range(10000)]", number=30)
0.040056066849501804
```

Reply:
edited into:  basicEventCombined = set(itertools.chain.from_iterable(*term))

Reply:
Fixed with suggested code